Some datastore fixes

sinkuu opened the pull request:
I tried using swiftds plugin(#3) on IPFS and found that it is not reporting correct disk usage and NumObjects is capped at 10000. This PR fixes these problems.
This PR also implements batched uploading & deletion.

This PR contains a breaking change (changes object names).

sinkuu committed:
datastore: implement batch operations

sinkuu committed:
datastore: implement DiskUsage

sinkuu committed:
datastore: list all objects correctly

`Object` returns only 10000 objects.

sinkuu committed:
datastore: implement GetSize

willscott commented on swift.go:
```diff
@@ -70,48 +79,74 @@ func (s *SwiftContainer) Has(k ds.Key) (bool, error) {
-	res := make([]dsq.Entry, len(objs[q.Offset:]))
-	for i, obj := range objs[q.Offset:] {
-		res[i] = dsq.Entry{Key: obj.Name}
+	names, err := s.conn.ObjectNamesAll(s.Container, &opts)
```
Is it possible to make use of `Marker` / `KeepMarker` to avoid having the query request all items from the first - Offset + Limit? Even if I have a relatively small Limit, If my query pages through all results, the final request would itself re-fetch the entire list of items, which seems pretty inefficient.

willscott reviewed (commented):
Comment on the call to Objects / ObjectNamesAll, but otherwise the collection of changes in this PR look good

sinkuu committed:
datastore: async Query

ObjectsNamesAll fetchs all the objects even when Limit option is set,
as it is the number of objects to fetch at once for each step in ObjectsWalk.
In this commit, we walks objects list ourself to prevent requesting all the items.

sinkuu committed:
datastore: query cache

willscott commented:
@sinkuu Thanks again for improving this datastore! think you'll have time to look into the `Marker` change?

sinkuu commented:
Sorry for being late. I rebased to master and added two commits:

* https://github.com/ipfs/go-ds-swift/pull/4/commits/e99074a24be05dab129d559eb681cee91a16a674 Implements object list walking ourself, because `ObjectNamesAll` API fetches all the item in the storage, even if `Limit` is set.
* https://github.com/ipfs/go-ds-swift/pull/4/commits/4b1d9b13e5e44c0e6e6ab1f92a221f4f4a5646a1 Caches the last fetched item and sets `Marker` option accordingly so that the adjacent queries (e.g. offset = 2, limit = 2 => offset = 4, limit = 2) are performed efficiently.

willscott reviewed (approved):
🙇‍♂️ Thanks! This should be quite a bit more efficient than it was!
